Source/Core/VideoCommon/OnScreenUI.cpp

```diff
@@ -0,0 +1,386 @@
+
+    io.Fonts->TexID = font_tex.get();
+
+    m_imgui_textures.push_back(std::move(font_tex));
+  }
+
+  if (!RecompileImGuiPipeline())
+    return false;
+
+  m_imgui_last_frame_time = Common::Timer::NowUs();
+  m_ready = true;
+  BeginImGuiFrameUnlocked(width, height);  // lock is already held
+
+  return true;
+}
+
+OnScreenUI::~OnScreenUI()
+{
+  std::unique_lock<std::mutex> imgui_lock(m_imgui_mutex);
+
+  ImGui::EndFrame();
+  ImPlot::DestroyContext();
+  ImGui::DestroyContext();
+}
+
+bool OnScreenUI::RecompileImGuiPipeline()
+{
+  if (g_presenter->GetBackbufferFormat() == AbstractTextureFormat::Undefined)
+  {
+    // No backbuffer (nogui) means no imgui rendering will happen
+    // Some backends don't like making pipelines with no render targets
+    return true;
+  }
+
+  std::unique_ptr<AbstractShader> vertex_shader = g_gfx->CreateShaderFromSource(
+      ShaderStage::Vertex, FramebufferShaderGen::GenerateImGuiVertexShader(),
+      "ImGui vertex shader");
+  std::unique_ptr<AbstractShader> pixel_shader = g_gfx->CreateShaderFromSource(
+      ShaderStage::Pixel, FramebufferShaderGen::GenerateImGuiPixelShader(), "ImGui pixel shader");
+  if (!vertex_shader || !pixel_shader)
+  {
+    PanicAlertFmt("Failed to compile ImGui shaders");
+    return false;
+  }
+
+  // GS is used to render the UI to both eyes in stereo modes.
+  std::unique_ptr<AbstractShader> geometry_shader;
+  if (g_gfx->UseGeometryShaderForUI())
+  {
+    geometry_shader = g_gfx->CreateShaderFromSource(
+        ShaderStage::Geometry, FramebufferShaderGen::GeneratePassthroughGeometryShader(1, 1),
+        "ImGui passthrough geometry shader");
+    if (!geometry_shader)
+    {
+      PanicAlertFmt("Failed to compile ImGui geometry shader");
+      return false;
+    }
+  }
+
+  AbstractPipelineConfig pconfig = {};
+  pconfig.vertex_format = m_imgui_vertex_format.get();
+  pconfig.vertex_shader = vertex_shader.get();
+  pconfig.geometry_shader = geometry_shader.get();
+  pconfig.pixel_shader = pixel_shader.get();
+  pconfig.rasterization_state = RenderState::GetNoCullRasterizationState(PrimitiveType::Triangles);
+  pconfig.depth_state = RenderState::GetNoDepthTestingDepthState();
+  pconfig.blending_state = RenderState::GetNoBlendingBlendState();
+  pconfig.blending_state.blendenable = true;
+  pconfig.blending_state.srcfactor = SrcBlendFactor::SrcAlpha;
+  pconfig.blending_state.dstfactor = DstBlendFactor::InvSrcAlpha;
+  pconfig.blending_state.srcfactoralpha = SrcBlendFactor::Zero;
+  pconfig.blending_state.dstfactoralpha = DstBlendFactor::One;
+  pconfig.framebuffer_state.color_texture_format = g_presenter->GetBackbufferFormat();
+  pconfig.framebuffer_state.depth_texture_format = AbstractTextureFormat::Undefined;
+  pconfig.framebuffer_state.samples = 1;
+  pconfig.framebuffer_state.per_sample_shading = false;
+  pconfig.usage = AbstractPipelineUsage::Utility;
+  m_imgui_pipeline = g_gfx->CreatePipeline(pconfig);
```

Comment:
This is a copy from `Renderer` but it surprises me.  I didn't realize the imgui shaders went away.